airbyte-integrations/connectors/source-mysql/src/test-integration/java/io/airbyte/integrations/source/mysql/CdcBinlogsMySqlSourceDatatypeTest.java

```diff
@@ -0,0 +1,171 @@
+import io.airbyte.db.Database;
+import io.airbyte.db.factory.DSLContextFactory;
+import io.airbyte.db.factory.DatabaseDriver;
+import io.airbyte.db.jdbc.JdbcUtils;
+import io.airbyte.integrations.standardtest.source.TestDataHolder;
+import io.airbyte.integrations.standardtest.source.TestDestinationEnv;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.JsonSchemaType;
+import java.util.List;
+import org.jooq.DSLContext;
+import org.jooq.SQLDialect;
+import org.testcontainers.containers.MySQLContainer;
+
+public class CdcBinlogsMySqlSourceDatatypeTest extends AbstractMySqlSourceDatatypeTest {
+
+  private DSLContext dslContext;
+  private JsonNode stateAfterFirstSync;
+
+  @Override
+  protected void tearDown(final TestDestinationEnv testEnv) {
+    dslContext.close();
+    container.close();
+  }
+
+  @Override
+  protected List<AirbyteMessage> runRead(ConfiguredAirbyteCatalog configuredCatalog) throws Exception {
+    if (stateAfterFirstSync == null) {
+      throw new RuntimeException("stateAfterFirstSync is null");
+    }
+    return super.runRead(configuredCatalog, stateAfterFirstSync);
+  }
+
+  @Override
+  protected void setupEnvironment(TestDestinationEnv environment) throws Exception {
+    final Database database = setupDatabase();
+    initTests();
+    for (final TestDataHolder test : testDataHolders) {
+      database.query(ctx -> {
+        ctx.fetch(test.getCreateSqlQuery());
+        return null;
+      });
+    }
+
+    final ConfiguredAirbyteStream dummyTableWithData = createDummyTableWithData(database);
+    final ConfiguredAirbyteCatalog catalog = getConfiguredCatalog();
+    catalog.getStreams().add(dummyTableWithData);
+
+    final List<AirbyteMessage> allMessages = super.runRead(catalog);
+    if (allMessages.size() != 2) {
+      throw new RuntimeException("First sync should only generate 2 records");
+    }
+    final List<AirbyteStateMessage> stateAfterFirstBatch = extractStateMessages(allMessages);
+    if (stateAfterFirstBatch == null || stateAfterFirstBatch.isEmpty()) {
+      throw new RuntimeException("stateAfterFirstBatch should not be null or empty");
+    }
+    stateAfterFirstSync = Jsons.jsonNode(stateAfterFirstBatch);
+    if (stateAfterFirstSync == null) {
+      throw new RuntimeException("stateAfterFirstSync should not be null");
+    }
+    for (final TestDataHolder test : testDataHolders) {
+      database.query(ctx -> {
+        test.getInsertSqlQueries().forEach(ctx::fetch);
+        return null;
+      });
+    }
+  }
+
+  @Override
+  protected Database setupDatabase() throws Exception {
+    container = new MySQLContainer<>("mysql:8.0");
+    container.start();
+
+    config = Jsons.jsonNode(ImmutableMap.builder()
+        .put(JdbcUtils.HOST_KEY, container.getHost())
```

Comment:
prefer HostPortResolver.resolveHost() and resolvePort() for the host/portname - otherwise this breaks on m1 macs (maybe? at the very least, george and I had issues with postgres tests which were fixed by switching to HostPortResolver)

(iirc the DSLContextFactory still needs to use container.getHost/getPort because that's running directly on the host laptop, not inside a container)